### PR TITLE
Update Apps and Clients registration API

### PR DESCRIPTION
**Pull request by yuliu-okta**
## Description:
- Update Apps and OAuth2 Client registration APIs
- EA OAuth2 and Client registration APIs

## Type of Pull Request:

- [x] Content (Documentation updates or typo-fixes)
- [ ] Blog Post
- [ ] Functional (CSS changes, JS updates, or layout modifications)

### Resolves:
* [OKTA-130073](https://oktainc.atlassian.net/browse/OKTA-130073)

## How Has This Been Tested?
- [ ] I have built this locally and verified that it does not break existing functionality.
- [ ] For functional changes, I have tested against Chrome, Firefox, and Safari, and mobile.
- [ ] For functional changes, I have added tests.

## Primary Reviewers:
@danielzeller-okta @mystiberry-okta 

- *Tag reviewer(s)*

- [Blog Post Google Doc](https://docs.google.com)
@cwu-okta @larsjohansen-okta @royalchan-okta @petermiller-okta

### CODE REVIEW

File: _source/_docs/api/resources/apps.md [hunk 1]
```diff
@@ -966,6 +966,200 @@ curl -v -X POST \
...
+* The `grant_types` and `response_types` values described above are partially orthogonal, as they refer to arguments passed to different
+    endpoints in the [OAuth 2.0 protocol](https://tools.ietf.org/html/rfc6749). However, they are related in that the `grant_types`
+    available to a client influence the `response_types` that the client is allowed to use, and vice versa. For instance, a `grant_types`
+    value that includes `authorization_code` implies a `response_types` value that includes `code`, as both values are defined as part of
+    the OAuth 2.0 authorization code grant.
+
+##### Request Example
+{:.api .api-request .api-request-example}
+
+> [Application](#application-model)'s "signOnMode" must be set to OPENID_CONNECT, the "name" field must be "oidc_client", and "label" field must be defined.
```
Comment (mystiberry-okta) [inline _source/_docs/api/resources/apps.md]: Can we have signOnMode, name, and label in tick marks `label` ?

File: _source/_docs/api/resources/apps.md [hunk 2]
```diff
@@ -4339,15 +4513,15 @@ Applications have the following properties:
...
[JSON HAL](http://tools.ietf.org/html/draft-kelly-json-hal-06)     | TRUE       | FALSE    | TRUE       |             |             |
 | _embedded        | embedded resources related to the app        | 
[JSON HAL](http://tools.ietf.org/html/draft-kelly-json-hal-06)     | TRUE       | FALSE    | TRUE       |             |             |
 | ---------------- | -------------------------------------------- | ------------------------------------------------------------------ | ---------- | -------- | ---------- | ----------- | ----------- |
 
 Property details
 
  * `id`, `created`, `lastUpdated`, `status`, `_links`, and `_embedded` are only available after an app is created.
- * `profile` is only available for OpenID Connect apps. See [Set Profile Property for OpenID Connect Apps](#set-profile-property-for-openid-connect-apps).
+ * `profile` is only available for OAuth2 Client apps. See [Profile Object](#profile-object).
```
Comment (mystiberry-okta) [inline _source/_docs/api/resources/apps.md]: OAuth 2.0 client

File: _source/_docs/api/resources/apps.md [hunk 3]
```diff
@@ -4494,6 +4670,7 @@ Specifies credentials and scheme for the application's `signOnMode`.
...
 | userName         | Shared username for app                                                                                        | String                                                    | TRUE     |                 | 1         | 100       |            |
 | password         | Shared password for app                                                                                        | [Password Object](#password-object)                       | TRUE     |                 |           |           |            |
+| oauthClient      | Credential for OAuth2 Client                                                                                   | [OAuth2 Credential Object](#oauth2-credential-object)   | False    |                 |           |           |            |
 |------------------+----------------------------------------------------------------------------------------------------------------+-----------------------------------------------------------+----------+-----------------+-----------+-----------+------------|
```
Comment (mystiberry-okta) [inline _source/_docs/api/resources/apps.md]: Change this link when you change the section title to OAuth 2.0 Credential Object.

File: _source/_docs/api/resources/apps.md [hunk 4]
```diff
@@ -4572,6 +4749,34 @@ Determines the [key](#application-key-credential-model) used for signing asserti
 }
 ~~~
 
+#### OAuth2 Credential Object
```
Comment (mystiberry-okta) [inline _source/_docs/api/resources/apps.md]: OAuth 2.0...

File: _source/_docs/api/resources/apps.md [hunk 5]
```diff
@@ -4572,6 +4749,34 @@ Determines the [key](#application-key-credential-model) used for signing asserti
...
+
+Determines how to authenticate the OAuth2 Client
```
Comment (mystiberry-okta) [inline _source/_docs/api/resources/apps.md]: OAuth 2.0 client

File: _source/_docs/api/resources/apps.md [hunk 6]
```diff
@@ -4572,6 +4749,34 @@ Determines the [key](#application-key-credential-model) used for signing asserti
...
+
+|----------------------------+----------------------------------------------------------------------------------+----------+----------|
+| Property                   | Description                                                                      | DataType | Nullable |
+| -------------------------- | ------------------------------------------------------------------------------------------- | -------- |
+| client_id                  | unique identifier for the OAuth2 Client application                              | String   | TRUE     |
```
Comment (mystiberry-okta) [inline _source/_docs/api/resources/apps.md]: Need capital letter in first letter of each Description.

File: _source/_docs/api/resources/apps.md [hunk 7]
```diff
@@ -4572,6 +4749,34 @@ Determines the [key](#application-key-credential-model) used for signing asserti
...
+| client_secret              | OAuth 2.0 client secret string                                                   | String   | TRUE     |
+| token_endpoint_auth_method | requested authentication method for the token endpoint                           | String   | FALSE    |
+|----------------------------+----------------------------------------------------------------------------------+----------+----------|
+
+* When creating a OAuth2 Client application, you can specify the `client_id`, or Okta will set it the same value as application id. Thereafter, the client_id is immutable.
```
Comment (mystiberry-okta) [inline _source/_docs/api/resources/apps.md]: an OAuth 2.0 client application

Comment (mystiberry-okta) [inline _source/_docs/api/resources/apps.md]: "as the application ID."

File: _source/_docs/api/resources/apps.md [hunk 8]
```diff
@@ -4572,6 +4749,34 @@ Determines the [key](#application-key-credential-model) used for signing asserti
...
+
+* If a `client_secret` is not provided on creation, and the `token_endpoint_auth_method` requires one Okta will generate a random `client_secret` for the client application. For update operation, the `client_secret` must be provided if the `token_endpoint_auth_method` requires one.
+
+* The `client_id` must consist of alphanumeric characters, the following special characters `$-_.+!*'(),`, must contain between 6 and 100 characters, inclusive, and must not be the reserved word `ALL_CLIENTS`. The `client_secret` must consist of printable characters, which are defined in [the OAuth 2.0 Spec](https://tools.ietf.org/html/rfc6749#appendix-A), and must contain between 14 and 100 characters, inclusive.
```
Comment (mystiberry-okta) [inline _source/_docs/api/resources/apps.md]: How about "The `client_id` must contain only alphanumeric characters or the following special characters: `$-).+!*'(),`. It must contain ....

Comment (mystiberry-okta) [inline _source/_docs/api/resources/apps.md]: Just to simplify the logic of the first clause and make it easier to read.

File: _source/_docs/api/resources/apps.md [hunk 9]
```diff
@@ -4720,6 +4925,18 @@ Basic           | urn:oasis:names:tc:SAML:2.0:attrname-format:basic
...
+{:.api .api-operation}
+
+Profile object is a container for any valid JSON schema that can be referenced from a request. For example, add an app manager contact email address, or define a whitelist of groups that you can then reference using the [Okta Expression `getFilteredGroups`](/reference/okta_expression_language/index.html#group-functions).
+
+Profile Requirements
+
+* The `profile` property is not encrypted, so don't store sensitive data in it.
+* The `profile` property doesn't limit the level of nesting in the JSON schema you created, but there is a practical size limit. We recommend a JSON schema size of 1 MB or less for best performance.
+
+> Profile Object is only avialible to OAuth2 Client applications.
```
Comment (mystiberry-okta) [inline _source/_docs/api/resources/apps.md]: OAuth 2.0 client

File: _source/_docs/api/resources/oauth-clients.md [hunk 10]
```diff
@@ -90,25 +64,22 @@ Client applications have the following properties:
...
-| _links                     | discoverable resources related to the app                         | [JSON HAL](http://tools.ietf.org/html/draft-kelly-json-hal-06)                               | TRUE     | FALSE  | TRUE     |
 |----------------------------+-------------------------------------------------------------------+----------------------------------------------------------------------------------------------+----------+--------+----------|
 
 Property Details
 
-* When creating a client application, you can specify the `client_id` or let Okta generate a value for it. Thereafter, the `client_id` is immutable.
-
-* The `client_secret` is only shown on the initial creation of a client application (and only if the `token_endpoint_auth_method` is one that requires a client secret).
-  It is never returned in a GET call. If a `client_secret` is not provided on creation and the `token_endpoint_auth_method` requires one Okta will generate a random `client_secret` for the client application.
+* When creating a client application, you can not specify the `client_id`, Okta uses application id for it. Thereafter, the `client_id` is immutable.
```
Comment (mystiberry-okta) [inline _source/_docs/api/resources/oauth-clients.md]: "cannot"

Comment (mystiberry-okta) [inline _source/_docs/api/resources/oauth-clients.md]: Also, Okta uses the application ID (or, Okta uses `app_id`?)

File: _source/_docs/api/resources/oauth-clients.md [hunk 11]
```diff
@@ -90,25 +64,22 @@ Client applications have the following properties:
...
 
-* The `client_id` must consist of alphanumeric characters, the following special characters `$-_.+!*'(),`, must contain between 6 and 100 characters, inclusive, and must not be the reserved word `ALL_CLIENTS`. The `client_secret` must consist of printable characters, which are defined in [the OAuth 2.0 Spec](https://tools.ietf.org/html/rfc6749#appendix-A), and must contain between 14 and 100 characters, inclusive.
+* The `client_secret` is only shown on the creation or update of a client application (and only if the `token_endpoint_auth_method` is one that requires a client secret).
+  It is never returned in a GET call. You can not specify the `client_secret` on creation and if the `token_endpoint_auth_method` requires one Okta will generate a random `client_secret` for the client application.
```
Comment (mystiberry-okta) [inline _source/_docs/api/resources/oauth-clients.md]: cannot

File: _source/_docs/api/resources/apps.md [hunk 12]
```diff
@@ -966,6 +966,200 @@ curl -v -X POST \
 }' "https://${org}.okta.com/api/v1/apps"
 ~~~
 
+#### Add OAuth2 Client Application
+{:.api .api-operation}
+
+Adds a OAuth2 Client application. This application is only available to the org that creates it.
```
Comment (yuliu-okta) [inline _source/_docs/api/resources/apps.md]: fixed

File: _source/_docs/api/resources/apps.md [hunk 13]
```diff
@@ -966,6 +966,200 @@ curl -v -X POST \
...
+
+> You must enable the OpenID Connect feature to create an OAuth2 Client. OpenId Connect is an {% api_lifecycle ea %} feature.
```
Comment (yuliu-okta) [inline _source/_docs/api/resources/apps.md]: fixed

File: _source/_docs/api/resources/apps.md [hunk 14]
```diff
@@ -966,6 +966,200 @@ curl -v -X POST \
...
+
+##### Credentials
+{:.api .api-request .api-request-params}
+
+Parameter             | Description                                                                                                       | DataType                                             | Nullable | Unique | Validation
+--------------------- | ----------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------- --| -------- | ----- | ----------------------------------------
+ client_id                  | unique identifier for the client application                                                                | String                                                 | TRUE     | TRUE  | TRUE     |
```
Comment (yuliu-okta) [inline _source/_docs/api/resources/apps.md]: fixed

File: _source/_docs/api/resources/apps.md [hunk 15]
```diff
@@ -966,6 +966,200 @@ curl -v -X POST \
...
+--------------------- | ---------------------------------------------------------------- | -------------------------------------------------------------------------------------------- | -------- | ------ | ----------------------------------------
+ client_uri           | URL string of a web page providing information about the client  | String                                                                                       | TRUE     | FALSE  | FALSE |
+ logo_uri             | URL string that references a logo for the client                 | String                                                                                       | TRUE     | FALSE  | FALSE |
+ application_type     | The type of client application                                   | `web`, `native`, `browser`, or `service`                                                     | TRUE     | TRUE   | TRUE  |
+ redirect_uris        | array of redirection URI strings for use in redirect-based flows | Array                                                                                        | TRUE     | FALSE  | FALSE |
```
Comment (yuliu-okta) [inline _source/_docs/api/resources/apps.md]: fixed

File: _source/_docs/api/resources/apps.md (hunk 1 above)
Comment (yuliu-okta) [inline _source/_docs/api/resources/apps.md]: changed

File: _source/_docs/api/resources/apps.md (hunk 2 above)
Comment (yuliu-okta) [inline _source/_docs/api/resources/apps.md]: fixed

File: _source/_docs/api/resources/apps.md (hunk 3 above)
Comment (yuliu-okta) [inline _source/_docs/api/resources/apps.md]: Change it to OAuth Credential Object

File: _source/_docs/api/resources/apps.md (hunk 4 above)
Comment (yuliu-okta) [inline _source/_docs/api/resources/apps.md]: OAuth Credential Object

File: _source/_docs/api/resources/apps.md (hunk 5 above)
Comment (yuliu-okta) [inline _source/_docs/api/resources/apps.md]: changed

File: _source/_docs/api/resources/apps.md (hunk 6 above)
Comment (yuliu-okta) [inline _source/_docs/api/resources/apps.md]: fixed

File: _source/_docs/api/resources/apps.md (hunk 7 above)
Comment (yuliu-okta) [inline _source/_docs/api/resources/apps.md]: fixed both

File: _source/_docs/api/resources/oauth-clients.md (hunk 10 above)
Comment (yuliu-okta) [inline _source/_docs/api/resources/oauth-clients.md]: changed

File: _source/_docs/api/resources/oauth-clients.md (hunk 11 above)
Comment (yuliu-okta) [inline _source/_docs/api/resources/oauth-clients.md]: changed

File: _source/_docs/api/resources/apps.md (hunk 8 above)
Comment (yuliu-okta) [inline _source/_docs/api/resources/apps.md]: fixed

File: _source/_docs/api/resources/apps.md (hunk 9 above)
Comment (yuliu-okta) [inline _source/_docs/api/resources/apps.md]: fixed

File: _source/_docs/platform-release-notes/platform-release-notes2017-20.md [hunk 16]
```diff
@@ -50,7 +50,7 @@ For example:
 * Add an app manager contact email address.
 * Use the profile to define a whitelist of groups that you can then reference and pass as claims using the [Okta Expression Language function `getFilteredGroups`](/reference/okta_expression_language/index.html#group-functions).
 
-For more information, see the [Apps API](/docs/api/resources/apps.html#set-profile-property-for-openid-connect-apps).
+For more information, see the [Apps API](/docs/api/resources/apps.html#profile-object).
```
Comment (larsjohansen-okta) [inline _source/_docs/platform-release-notes/platform-release-notes2017-20.md]: Can we change the release notes that's already gone out?

File: _source/_docs/api/resources/apps.md [hunk 17]
```diff
@@ -966,6 +966,200 @@ curl -v -X POST \
...
+
+##### Settings
+{:.api .api-request .api-request-params}
+
+Parameter             | Description                                                      | DataType                                                                                     | Nullable | Unique | Validation
+--------------------- | ---------------------------------------------------------------- | -------------------------------------------------------------------------------------------- | -------- | ------ | ----------------------------------------
+ client_uri           | URL string of a web page providing information about the client  | String                                                                                       | TRUE     | FALSE  | FALSE |
+ logo_uri             | URL string that references a logo for the client                 | String                                                                                       | TRUE     | FALSE  | FALSE |
+ application_type     | The type of client application                                   | `web`, `native`, `browser`, or `service`                                                     | TRUE     | TRUE   | TRUE  |
```
Comment (danielzeller-okta) [inline _source/_docs/api/resources/apps.md]: application_type isn't unique

File: _source/_docs/api/resources/apps.md [hunk 18]
```diff
@@ -966,6 +966,200 @@ curl -v -X POST \
...
+ redirect_uris        | Array of redirection URI strings for use in redirect-based flows | Array                                                                                        | TRUE     | FALSE  | FALSE |
```
Comment (danielzeller-okta) [inline _source/_docs/api/resources/apps.md]: we do validate the redirect_uris

File: _source/_docs/api/resources/apps.md [hunk 19]
```diff
@@ -966,6 +966,200 @@ curl -v -X POST \
...
+ response_types       | Array of OAuth 2.0 response type strings                         | Array of `code`, `token`, `id_token`                                                         | TRUE     | FALSE  | FALSE |
+ grant_types          | Array of OAuth 2.0 grant type strings                            | Array of `authorization_code`, `implicit`, `password`, `refresh_token`, `client_credentials` | FALSE    | FALSE  | FALSE |
```
Comment (danielzeller-okta) [inline _source/_docs/api/resources/apps.md]: we do validate the response_type/grant_type/application_type combo, not sure if that's something we want to call out here though

File: _source/_docs/api/resources/oauth-clients.md [hunk 20]
```diff
@@ -90,25 +64,22 @@ Client applications have the following properties:
...
-| _links                     | discoverable resources related to the app                         | [JSON HAL](http://tools.ietf.org/html/draft-kelly-json-hal-06)                               | TRUE     | FALSE  | TRUE     |
 |----------------------------+-------------------------------------------------------------------+----------------------------------------------------------------------------------------------+----------+--------+----------|
 
 Property Details
 
-* When creating a client application, you can specify the `client_id` or let Okta generate a value for it. Thereafter, the `client_id` is immutable.
-
-* The `client_secret` is only shown on the initial creation of a client application (and only if the `token_endpoint_auth_method` is one that requires a client secret).
-  It is never returned in a GET call. If a `client_secret` is not provided on creation and the `token_endpoint_auth_method` requires one Okta will generate a random `client_secret` for the client application.
+* When creating a client application, you cannot specify the `client_id`, Okta uses application ID for it. Thereafter, the `client_id` is immutable.
```
Comment (danielzeller-okta) [inline _source/_docs/api/resources/oauth-clients.md]: seems to me like `Thereafter` is redundant, as the client_id is ALWAYS immutable on this endpoint

Comment (danielzeller-okta) [inline _source/_docs/api/resources/oauth-clients.md]: Should we call out here that you CAN specify the client id and secret on the apps api?

File: _source/_docs/platform-release-notes/platform-release-notes2017-20.md (hunk 16 above)
Comment (yuliu-okta) [inline _source/_docs/platform-release-notes/platform-release-notes2017-20.md]: This is requested by @mystiberry-okta

File: _source/_docs/api/resources/oauth-clients.md (hunk 20 above)
Comment (yuliu-okta) [inline _source/_docs/api/resources/oauth-clients.md]: added

File: _source/_docs/api/resources/apps.md (hunk 19 above)
Comment (yuliu-okta) [inline _source/_docs/api/resources/apps.md]: change them all to validated

File: _source/_docs/api/resources/apps.md (hunk 18 above)
Comment (yuliu-okta) [inline _source/_docs/api/resources/apps.md]: yes

File: _source/_docs/api/resources/apps.md (hunk 17 above)
Comment (yuliu-okta) [inline _source/_docs/api/resources/apps.md]: fixed

File: _source/_docs/platform-release-notes/platform-release-notes2017-20.md (hunk 16 above)
Comment (mystiberry-okta) [inline _source/_docs/platform-release-notes/platform-release-notes2017-20.md]: He is changing the release note that's already gone out, so I think we're good?

**Comment by mystiberry-okta** [comment]
@yuliu-okta Can you resolve the conflicts so I can merge this? Thanks!

**Comment by mystiberry-okta** [comment]
I see some more link breaks in other files. I will fix. But hoping YL can do the conflict resolution.

**Pull request merged**